File: server/helpers/gitHelpers.js
// helpers/gitHelpers.js

/**
 * Clones a GitHub repository to a target directory.
 * @param {string} repoUrl - The GitHub repository URL.
 * @param {string} targetDir - The absolute path where the repo should be cloned.
 * @param {string} [token] - Optional GitHub Personal Access Token for private repos.
 * @param {function} execPromise - Promisified version of Node.js exec function. (It executes shell commands and returns a promise)
 */
async function cloneRepo(repoUrl, targetDir, token, execPromise) {
    // If a token is provided, insert it into the repo URL (for HTTPS)
    let cloneUrl = repoUrl;
    if (token && repoUrl.startsWith('https://')) {
      // Insert token after https:// (be mindful of encoding or special characters)
      cloneUrl = repoUrl.replace(
        'https://',
        `https://${token}@`
      );
    } else {
      throw new Error('Invalid GitHub repository URL');
    }
  
    const command = `git clone ${cloneUrl} ${targetDir}`;
    console.log('Cloning repository with command:', command);
    try {
      await execPromise(command);
      return targetDir;
    } catch (error) {
      throw new Error(`Error cloning repository: ${error.message}`);
    }
  }

/**
 * Sanitizes a name for use as the target directory for a 'git clone' command.
 * @param {string} name - The name to sanitize.
 * @returns {string} Sanitized name.
 */
function sanitizeName(name) {
  return name.replace(/[^a-z0-9]/gi, '_');
}

module.exports = {
  cloneRepo,
  sanitizeName
};

File: server-python/helpers/git_helpers.py
import os
import subprocess
import shutil
from pathlib import Path

def clone_repo(repo_url, target_dir, token=None):
    """
    Clone a git repository to a target directory
    
    Args:
        repo_url (str): The URL of the repository to clone
        target_dir (str): Path to the directory where to clone
        token (str, optional): GitHub token for private repos
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        # Create target directory if it doesn't exist
        Path(target_dir).mkdir(parents=True, exist_ok=True)
        
        # Prepare git clone command
        if token and repo_url.startswith('https://'):
            # Insert token into URL for private repos
            parts = repo_url.split('://')
            repo_url = f"{parts[0]}://{token}@{parts[1]}"
        
        # Execute git clone
        subprocess.run(
            ['git', 'clone', repo_url, target_dir],
            check=True,
            capture_output=True,
            text=True
        )
        
        return True
    except subprocess.CalledProcessError as e:
        print(f"Git clone error: {e.stderr}")
        # Clean up directory if clone failed
        if os.path.exists(target_dir):
            shutil.rmtree(target_dir)
        return False
    except Exception as e:
        print(f"Error cloning repository: {str(e)}")
        # Clean up directory if clone failed
        if os.path.exists(target_dir):
            shutil.rmtree(target_dir)
        return False 